common/ratelimit.go

```diff
@@ -46,14 +46,23 @@ type RateBucketParams struct {
 	LastRequestTime time.Time
 }
 
-func GetClientAddress(ctx context.Context, header string) (string, error) {
-	if header != "" {
+// GetClientAddress returns the client address from the context. If the header is not empty, it will
+// take the ip address located at the `numProxies“ position from the end of the header. If the ip address cannot be
+// found in the header, it will use the connection ip if `alloweDirectionConnection` is true. Otherwise, it will return
+// an error.
+func GetClientAddress(ctx context.Context, header string, numProxies int, allowDirectConnectionFallback bool) (string, error) {
+
+	if header != "" && numProxies > 0 {
 		md, ok := metadata.FromIncomingContext(ctx)
-		if !ok || len(md.Get(header)) == 0 {
-			return "", fmt.Errorf("failed to get ip from header")
+		if ok && len(md.Get(header)) > 0 {
+			parts := splitHeader(md.Get(header))
+			if len(parts) >= numProxies {
```

Comment:
Should this error out if numProxies is greater than len(parts)? It's basically malformed parameters to this function

Reply:
It could be a direct connection. If `allowDirectConnection` is true, it will use the direct connection in this case. Otherwise it will error out.